llama-index-core/llama_index/core/instrumentation/dispatcher.py

```diff
@@ -261,20 +263,53 @@ def wrapper(func: Callable, instance: Any, args: list, kwargs: dict) -> Any:
                 parent_id=parent_id,
                 tags=tags,
             )
+
+            def handle_future_result(future, span_id, bound_args, instance):
+                try:
+                    result = future.result()
+                    self.span_exit(
+                        id_=span_id,
+                        bound_args=bound_args,
+                        instance=instance,
+                        result=result,
+                    )
+                    return result
+                except BaseException as e:
+                    self.event(SpanDropEvent(span_id=span_id, err_str=str(e)))
+                    self.span_drop(
+                        id_=span_id, bound_args=bound_args, instance=instance, err=e
+                    )
+                    raise
+                finally:
+                    active_span_id.reset(token)
+
             try:
                 result = func(*args, **kwargs)
+                if isinstance(result, asyncio.Future):
+                    # If the result is a Future, wrap it
+                    new_future = asyncio.ensure_future(result)
+                    new_future.add_done_callback(
```

Comment:
This makes sense to me! Wonder where that Context is being spun up in the arize error you shared

Reply:
Thank you for the quick fix @logan-markewich! Just wanted to callout that the same error occurs with MLflow tracing (while the trace itself is logged correctly).

From my observation, I think the error is because of trying to reset `ContextVar` by a token that is created in the main thread, while resetting it in the async task - Arize and MLflow works fine nevertheless because neither relies on `active_span_id`. I think setting `active_span_id` inside `handle_future_result` should fix it (and it makes more sense that the async task manages its own active span).

Reply:
@B-Step62 I don't think I can move the `active_span_id` into `handle_future_result`, since its needed for `span_enter`, and `handle_future_result` only runs once the span ends 🤔

Reply:
@logan-markewich I see. I think the challenge here is we are trying to put both sync and async execution into a single `active_span`. In reality, they are executed in separate time.

1. First we enter the wrapper.
2. Then we create a future object and exit the wrapper immediately.
3. After a while, the future is scheduled for execution.
4. Finally, the task is completed.

There is some time Between 2 and 3, depending on when users call await. Currently, we are trying to create a single span that stays "active" all the way from 1 to 4. However, between 2 and 3, another sync operation should be able to take over the "active" span.

Therefore, I think we should manage active span like this
1. First we enter the wrapper. -> **Create a span and set to active**
2. Then we create a future object and exit the wrapper immediately. -> **Reset active span**
3. After a while, the future is scheduled for execution. -> **Set span to active**. This lives in the separate context so should not conflict with the "active" span in the main context, while any spans created within the async task will recognize this as the parent span.
4. Finally, the task is completed. -> **Reset active span**

How does this sound? This should be correct in theory, but not 100% sure if this works.